Allow builds to be cross compiled

btharper opened the pull request:
Adds cross compilation toolchain to dapper process when not on arm64 (currently packages are available on amd64, i386, and ppc64el from debian).

Also bumps version to 4.19 to match current branch of Raspbian.

Kernel built on amd64 tested and boots from branches 4.14.y, 4.19.y, and 5.2.y

With modifications to defconfig, dtb, firmware, and config.txt, a pi4 system can be produced on 4.19.y branch (5.2.y untested).

btharper committed:
Add cross compilation support for amd64 hosts

btharper committed:
Limit files added to /boot

New versions add files for the Raspberry Pi 4 that when added exceed the 30MB
boot partition for RancherOS. The plain set boots standard config, and the
"cd" versions are cutdown versions for gpu_mem=16 that disables some GPU
features. Removes the debug (db) and extended (x) versions, debug versions
are less common, but the x versisons are required for camera use.

btharper committed:
Allow for more kernel versions and only build required parts

btharper committed:
Bump version to match current Raspbian release

btharper committed:
Remove KERNEL_VERSION from Dockerfile.dapper and detect during build

Jason-ZW commented on Dockerfile.dapper:
```diff
@@ -27,8 +31,7 @@ ARG DAPPER_HOST_ARCH
 ARG KERNEL_URL=https://github.com/raspberrypi/linux.git
-ARG KERNEL_BRANCH=rpi-4.14.y
-ARG KERNEL_VERSION=4.14.114
+ARG KERNEL_BRANCH=rpi-4.19.y
```
Please do not upgrade the kernel version although this is a meaningful thing, because `rancher/os` expects to upgrade the kernel after a series of stability tests.

btharper replied:
Version bump was in service to getting things working on the Pi4, but that's probably out of scope for this PR anyway. Is leaving the minor version floating and just pinning the branch okay?

Jason-ZW reviewed (changes requested):
@btharper You also need to rebase the code. Others LGTM. After that, we will test. Thanks.

Jason-ZW commented:
#1

btharper committed:
Remove version bump from PR

Jason-ZW reviewed (approved):
LGTM

btharper commented:
Is the rebase just for linear history? I don't see anything conflicting across branches.

Jason-ZW commented:
There's nothing more to do thank you

Jason-ZW commented on Dockerfile.dapper:
```diff
@@ -18,6 +18,10 @@ RUN apt-get update \
 		locales \
+	&& if [ "${DAPPER_HOST_ARCH}" != "arm64" ]; then \
+		DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+		gcc-aarch64-linux-gnu; \
```
This line occurs an error. I think this is the right way `gcc-*-aarch64-linux-gnu`.

btharper replied:
I don't receive an error while running make using this file. What host arch are you using to build? and what error are you receiving?

[gcc-aarch64-linux-gnu](https://packages.debian.org/stretch/gcc-aarch64-linux-gnu) is the general package for Debian 9. It depends on the specific package gcc-6-aarch64-linux-gn. (gcc-8 in Debian 10, gcc-9 in Debian 11).

Jason-ZW replied:
aws Ubuntu-18.04 aarch64 aarch64 aarch64 GNU/Linux
gcc version 7.4.0
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic

btharper replied:
DAPPER_HOST_ARCH value is empty until the line `ARG DAPPER_HOST_ARCH` gives it a value. I've added a commit to address this. I'm currently testing on a x86_64 desktop and a RPi 3 but don't have results yet.

btharper replied:
Things look like they're working on both amd64 and arm64. Build on x86_64 completed without errors. RPi3 build made it past the line that was failing previously, but will not be finished quickly.

Jason-ZW replied:
Waiting for your RPi3 build result, after that, I will merge it.

btharper replied:
I'm letting it run overnight. So far, after 40 minutes, it hasn't even made it to the Dockerfile entrypoint. Kernel compile is also expected to take 3-5 hours, but I'll let you know once I have a result.

btharper replied:
Build finished. No errors on the last page of the scroll buffer.

Jason-ZW reviewed (changes requested):
Please modify the PR, see review comments.

Jason-ZW commented on Dockerfile.dapper:
```diff
@@ -18,6 +18,10 @@ RUN apt-get update \
+	&& if [ "${DAPPER_HOST_ARCH}" != "arm64" ]; then \
+		DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
+		gcc-aarch64-linux-gnu; \
+		fi \
```
Please formatting indentation

btharper replied:
Just indenting the continued command? Like this?
```
        && if [ "${DAPPER_HOST_ARCH}" != "arm64" ]; then \
                DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
                        gcc-aarch64-linux-gnu; \
                fi \
```

Jason-ZW replied:
> Just indenting the continued command? Like this?
> 
> ```
>         && if [ "${DAPPER_HOST_ARCH}" != "arm64" ]; then \
>                 DEBIAN_FRONTEND=noninteractive apt-get install -y --no-install-recommends \
>                         gcc-aarch64-linux-gnu; \
>                 fi \
> ```

```
&& if ...
          ....
      fi \
```

Jason-ZW replied:
my mistake that's fine

Jason-ZW commented on Dockerfile.dapper:
```diff
@@ -18,6 +18,10 @@ RUN apt-get update \
 		libncurses5-dev \
 		libssl-dev \
 		locales \
+	&& if [ "${DAPPER_HOST_ARCH}" != "arm64" ]; then \
```
aarch64 also needs to be considered

btharper replied:
aarch64 and arm64 are different ways to refer to the same thing. DAPPER_HOST_ARCH is populated with the Docker server arch or the contents of GOARCH.

On my RancherOS raspberry pi I get:
`docker version -f "{{.Server.Arch}}"` gives arm64
`uname -m` gives aarch64

btharper committed:
Define ARG variables before using them

Allows reference to DAPPER_HOST_ARCH to be non-empty when used to determine if ARCH=arm64
